src/main/java/frc/robot/subsystems/swerve/Module.java

```diff
@@ -0,0 +1,132 @@
+// Copyright 2023-2025 FRC 8033
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// version 3 as published by the Free Software Foundation or
+// available in the root directory of this project.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+package frc.robot.subsystems.swerve;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.SwerveModulePosition;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import org.littletonrobotics.junction.Logger;
+
+/** Wrapper around ModuleIO and ModuleIOInputs to organize module-level functionality. */
+public class Module {
+  // Represents per-module constants
+  public record ModuleConstants(
+      int id, String prefix, int driveID, int turnID, int cancoderID, Rotation2d cancoderOffset) {}
+
+  private final ModuleIO io;
+  private final ModuleIOInputsAutoLogged inputs = new ModuleIOInputsAutoLogged();
+
+  public Module(final ModuleIO io) {
+    this.io = io;
+  }
+
+  /** Update inputs without running the rest of the periodic logic. */
+  public void updateInputs() {
+    io.updateInputs(inputs);
+  }
+
+  public void periodic() {
+    Logger.processInputs(String.format("Swerve/%s Module", inputs.prefix), inputs);
+  }
+
+  /** Runs the module closed loop with the specified setpoint state. Returns the optimized state. */
+  public SwerveModuleState runSetpoint(SwerveModuleState state) {
+    return runSetpoint(state, 0.0, 0.0);
+  }
+
+  /** Runs the module closed loop with the specified setpoint state. Returns the optimized state. */
+  public SwerveModuleState runSetpoint(
+      SwerveModuleState state, double forceXNewtons, double forceYNewtons) {
```

Comment:
can i ask why/how this is based on force now

Reply:
this is to use Choreo's force calculations to help with auto path following. the previous acceleration was quite jank and didnt work very well, this should align better with the sorts of feedforwards we want to use it for.